File: web/src/session/turnMentions.ts
export type TurnMention = { kind: 'expert' | 'skill' | 'member'; id: string; name: string }

const isSpace = (ch: string) => /\s/.test(ch)

export function parseTurnMentions(text: string): TurnMention[] {
  const out: TurnMention[] = []
  const seen = new Set<string>()
  const add = (kind: TurnMention['kind'], id: string, name: string) => {
    const next: TurnMention = { kind, id: id.trim(), name: name.trim() }
    if (!next.kind || (!next.id && !next.name)) return
    const key = `${next.kind}\0${next.id}\0${next.name}`
    if (seen.has(key)) return
    seen.add(key)
    out.push(next)
  }
  scanPrefixedRefs(text, '[引用专家 ', 'expert', add)
  scanPrefixedRefs(text, '[引用技能 ', 'skill', add)
  for (const name of scanAtNames(text)) add('member', '', name)
  return out
}

function scanPrefixedRefs(text: string, prefix: string, kind: TurnMention['kind'], add: (kind: TurnMention['kind'], id: string, name: string) => void) {
  let rest = text
  for (;;) {
    const i = rest.indexOf(prefix)
    if (i < 0) return
    rest = rest.slice(i + prefix.length)
    const bar = rest.indexOf('|')
    const end = rest.indexOf(']')
    if (bar < 0 || end < 0 || bar >= end) continue
    add(kind, rest.slice(bar + 1, end), rest.slice(0, bar))
    rest = rest.slice(end + 1)
  }
}

function scanAtNames(text: string): string[] {
  const names: string[] = []
  const runes = [...text]
  for (let i = 0; i < runes.length; i++) {
    if (runes[i] !== '@') continue
    if (i > 0 && !isSpace(runes[i - 1]!)) continue
    let j = i + 1
    while (j < runes.length && !isSpace(runes[j]!) && runes[j] !== '@') j++
    if (j === i + 1) continue
    names.push(runes.slice(i + 1, j).join(''))
    i = j - 1
  }
  return names
}

export function extractExpertRefNames(text: string): string[] {
  const names: string[] = []
  const seen = new Set<string>()
  for (const m of parseTurnMentions(text)) {
    if (m.kind !== 'expert' || !m.name || seen.has(m.name)) continue
    seen.add(m.name)
    names.push(m.name)
  }
  return names
}

File: internal/app/turn_mentions.go
package app

import (
	"strings"
	"unicode"
)

func ParseTurnMentions(text string) []TurnMention {
	var out []TurnMention
	seen := map[string]bool{}
	add := func(kind, id, name string) {
		kind, id, name = strings.TrimSpace(kind), strings.TrimSpace(id), strings.TrimSpace(name)
		if kind == "" || (id == "" && name == "") {
			return
		}
		key := kind + "\x00" + id + "\x00" + name
		if seen[key] {
			return
		}
		seen[key] = true
		out = append(out, TurnMention{Kind: kind, ID: id, Name: name})
	}
	scanPrefixedRefs(text, "[引用专家 ", "expert", add)
	scanPrefixedRefs(text, "[引用技能 ", "skill", add)
	for _, name := range scanAtNames(text) {
		add("member", "", name)
	}
	return out
}

func scanPrefixedRefs(text, prefix, kind string, add func(kind, id, name string)) {
	rest := text
	for {
		i := strings.Index(rest, prefix)
		if i < 0 {
			return
		}
		rest = rest[i+len(prefix):]
		bar := strings.IndexByte(rest, '|')
		end := strings.IndexByte(rest, ']')
		if bar < 0 || end < 0 || bar >= end {
			continue
		}
		add(kind, strings.TrimSpace(rest[bar+1:end]), strings.TrimSpace(rest[:bar]))
		if end+1 >= len(rest) {
			return
		}
		rest = rest[end+1:]
	}
}

func scanAtNames(text string) []string {
	var names []string
	runes := []rune(text)
	for i := 0; i < len(runes); i++ {
		if runes[i] != '@' {
			continue
		}
		if i > 0 && !unicode.IsSpace(runes[i-1]) {
			continue
		}
		j := i + 1
		for j < len(runes) && !unicode.IsSpace(runes[j]) && runes[j] != '@' {
			j++
		}
		if j == i+1 {
			continue
		}
		names = append(names, string(runes[i+1:j]))
		i = j - 1
	}
	return names
}

func extractExpertRefNames(text string) []string {
	var names []string
	seen := map[string]bool{}
	for _, m := range ParseTurnMentions(text) {
		if m.Kind != "expert" || m.Name == "" || seen[m.Name] {
			continue
		}
		seen[m.Name] = true
		names = append(names, m.Name)
	}
	return names
}
